### PR TITLE
feat: add option to choose gas price from different platforms

### PR DESCRIPTION
**Pull request by guanbinrui**
### Summary

There are many `gasPrice` providers on the market. This PR is going to support  [gasnow](https://gasnow.org) and [gasstation](https://ethgasstation.info). And a new option will be added for switching between different platforms on the settings panel. On the front-end, a shared component will be used to modify the current gas price.

#### `GasPrice`

The `GasPrice` constrains the return type of different price providers.

#### `GasPriceProviderType`

All of the supported providers list in one enum. This enum should be extended if a new provider to be added in the future.

#### `<SelectGasPriceDialog />`

The front-end surface used to change the current gas price stored in the background.

#### `currentGasPriceSettings`

Tracks the current selected gas price in WEI.

### Screenshots

gas

### CODE REVIEW

File: packages/maskbook/src/utils/flags.ts [hunk 1]
```diff
@@ -29,6 +29,7 @@ export const Flags = {
     requires_injected_script_run_directly: process.env.target === 'firefox',
     // TODO: document why it enabled on app
     support_eth_network_switch: appOnly || betaOrInsiderOnly,
+    support_gas_price_server_switch: appOnly || betaOrInsiderOnly,
```
Comment (Jack-Works) [inline packages/maskbook/src/utils/flags.ts:32]: why it enabled on app

File: packages/maskbook/src/plugins/Wallet/types.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import type { CurrencyType } from '../../web3/types'
+
+export interface GasPrice {
+    title: string
+    description?: string
+    gasPrice: string
+    estimated?: {
+        [key in CurrencyType]: string
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Wallet/types.ts:8]: Record<keyof CurrencyType, string> ?

**Review by Jack-Works — APPROVED** [review]

**Review by Jack-Works — DISMISSED** [review]

File: packages/maskbook/src/web3/hooks/useGasPrice.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import { useValueRef } from '../../utils/hooks/useValueRef'
+import { getGasPrices } from '../apis'
+
+export function useGasPrice() {
+    const selectedGasPriceServer = useValueRef(currentSelectedGasPriceServerSettings)
+
+    return useAsyncRetry(async () => {
+        const gasPrices = await getGasPrices(selectedGasPriceServer)
+        return gasPrices.reverse()
+    }, [])
```
Comment (Jack-Works) [inline packages/maskbook/src/web3/hooks/useGasPrice.ts:12]: Please add the dep correctly

Comment (lelenei) [reply packages/maskbook/src/web3/hooks/useGasPrice.ts]: replace useAsync

File: packages/maskbook/src/web3/UI/EthereumGasButton.tsx [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+import SpacedButtonGroup from '../../extension/options-page/DashboardComponents/SpacedButtonGroup'
+import BigNumber from 'bignumber.js'
+import { useTransakGetPriceFroETH } from '../hooks/useTransakGetPriceFroETH'
+
+interface CalcETHAmountProps {
+    amount: string
+}
+function CalcETHAmount(props: CalcETHAmountProps) {
+    const { amount } = props
+    const { loading, value: fiat } = useTransakGetPriceFroETH(amount)
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/UI/EthereumGasButton.tsx:31]:
```suggestion
    const { loading, value: fiat } = useTransakGetPriceForETH(amount)
```

File: packages/maskbook/src/web3/UI/EthereumGasButton.tsx [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+            backgroundColor: 'grey',
+        },
+        price: {
+            paddingBottom: theme.spacing(1),
+            flexDirection: 'row',
+            justifyContent: 'space-between',
+            display: 'flex',
+            alignItems: 'center',
+            height: 34,
+            '& > *: last-child': {
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/UI/EthereumGasButton.tsx:76]: ?

File: packages/maskbook/src/web3/UI/EthereumGasButton.tsx [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+                flex: 1,
+                paddingRight: theme.spacing(1),
+            },
+        },
+        eth: {
+            marginTop: theme.spacing(1),
+        },
+        text: {
+            borderRadius: 5,
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/UI/EthereumGasButton.tsx:85]: `theme.shape.borderRadius`

File: packages/maskbook/src/web3/UI/EthereumGasButton.tsx [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+
+interface GasPriceItemProps {
+    gasPrice: GasPrice
+    primary?: string
+    secondary?: string
+    active?: boolean
+    onChange?: (gasPrice: GasPrice) => void
+}
+function CalcETH(amount: string) {
+    return new BigNumber(amount).multipliedBy('0.00000002').multipliedBy(GAS_LIMIT).toFixed()
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/UI/EthereumGasButton.tsx:102]: What does this mean? Is `0.00000002` a magical number?

File: packages/maskbook/src/web3/UI/EthereumGasButton.tsx [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+            setFocuse(true)
+        }
+    }, [gasPrice, onChange])
+
+    return (
+        <div
+            className={classes.root}
+            style={{
+                backgroundColor: hover ? secondary : 'inherit',
+                opacity: hover ? '0.6' : '1',
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/UI/EthereumGasButton.tsx:151]: Maybe you need `<Card />` with `<ButtonBase />` to make it behave like a button. I think `<Provider />` is a good example of this.

File: packages/maskbook/src/web3/UI/EthereumStatusBar.tsx [hunk 9]
```diff
@@ -1,13 +1,20 @@
...
 const useStyles = makeStyles((theme: Theme) =>
     createStyles({
         root: {
             display: 'flex',
             alignItems: 'center',
             justifyContent: 'flex-end',
+            '& > * ': {
+                marginRight: theme.spacing(1),
+            },
+            '& > * : last-child': {
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/UI/EthereumStatusBar.tsx:15]: I think too much global selector is an anti-pattern. Please pay attention to it.

File: packages/maskbook/src/web3/UI/EthereumGasButton.tsx [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+function CalcETH(amount: string) {
+    return new BigNumber(amount).multipliedBy('0.00000002').multipliedBy(GAS_LIMIT).toFixed()
+}
+
+function GasPriceItem(props: GasPriceItemProps) {
+    const { gasPrice, primary = 'green', secondary = 'grey', active = false, onChange } = props
+    const [hover, setHover] = useState(false)
+    const [amountForUI, setAmountForUI] = useState(gasPrice.gasPrice)
+    const classes = useGasPriceItemStyles({ hover })
+    const [eth, setETH] = useState(CalcETH(gasPrice.gasPrice))
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/UI/EthereumGasButton.tsx:110]: Why not move the ether calculating logic to the `<CalcETH  />`?

File: packages/maskbook/src/web3/UI/EthereumGasButton.tsx (hunk 6 above)
Comment (lelenei) [reply packages/maskbook/src/web3/UI/EthereumGasButton.tsx:85]: theme.shape.borderRadius is large

File: packages/maskbook/src/web3/UI/EthereumGasButton.tsx [hunk 11]
```diff
@@ -24,40 +25,28 @@ import BigNumber from 'bignumber.js'
...
-            border: '1px solid grey',
+            textAlign: 'left',
         },
+
         title: {
             display: 'flex',
             flexDirection: 'row',
             justifyContent: 'space-between',
-            alignItems: 'center',
+            alignItems: 'cener',
```
Comment (nuanyang233) [inline packages/maskbook/src/web3/UI/EthereumGasButton.tsx:49]: This is wrong...

File: packages/maskbook/src/web3/hooks/useTransakGetPriceForETH.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
+import { useState } from 'react'
+import { useAsync } from 'react-use'
+import { currentGasPriceEthToUSD } from '../../settings/settings'
+import { getTransakGetPriceForETH } from '../apis/getTransakGetPriceForETH'
+
+const timeout = 300
+let getTransakGetPriceForETH_Time = 0
+
+export function useTransakGetPriceForETH() {
+    console.log('time: ', getTransakGetPriceForETH_Time)
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/hooks/useTransakGetPriceForETH.ts:10]: Remove logs for debugging.

File: packages/maskbook/src/web3/constants/index.ts [hunk 13]
```diff
@@ -1,5 +1,8 @@
 import { ChainId } from '../types'
 
+export const GAS_CUSTOM_WAIT = 5
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/constants/index.ts:3]: Please comment the unit out.

File: packages/maskbook/src/web3/UI/EthereumGasMenuButton.tsx [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+                {...ButtonProps}
+                endIcon={loading ? null : <ArrowDropDownIcon fontSize="small" />}
+                aria-haspopup="true"
+                onClick={handleClickListItem}>
+                {loading ? (
+                    <CircularProgress size="1.5rem" />
+                ) : gasPrices && gasPrices.length > 0 ? (
+                    `${gasPrices?.[selectedIndex].gasPrice} | ${gasPrices?.[selectedIndex].title}`
+                ) : (
+                    'No Data'
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/UI/EthereumGasMenuButton.tsx:57]: i18n

File: packages/maskbook/src/web3/UI/EthereumStatusBar.tsx (hunk 9 above)
Comment (guanbinrui) [reply packages/maskbook/src/web3/UI/EthereumStatusBar.tsx:15]: Acutally, `& > * : last-child` it an invalid selector.

File: packages/maskbook/src/web3/apis/gasnow.ts [hunk 15]
```diff
@@ -0,0 +1,57 @@
+import BigNumber from 'bignumber.js'
+import type { GasPrice } from '../../plugins/Wallet/types'
+
+const GAS_NOW_API = 'https://www.gasnow.org/api/v3/gas/price'
+const WAIT_TIME = 8
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/apis/gasnow.ts:5]: Please leave a comment about the unit.

File: packages/maskbook/src/web3/apis/gasnow.ts [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+            data: GasNowData
+        }>
+    } catch (e) {
+        return {
+            data: null,
+        }
+    }
+}
+export async function getGasPrice(): Promise<GasPrice[]> {
+    const timestamp = Date.now() / 1000
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/apis/gasnow.ts:37]: Attention! `timestamp` is a float number.

File: packages/maskbook/src/web3/apis/index.ts [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+                    wait: 15,
+                    gasPrice: gasnow.gweiToGwei((await getGasPrice(chainId)) || '0'),
+                },
+            ]
+        case GasPriceProviderType.GasNow:
+            return gasnow.getGasPrice()
+        case GasPriceProviderType.GasStation:
+            return gasstation.getGasPrice()
+        default:
+            return unreachable(provider)
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/apis/index.ts:23]: Remove `retrurn`

File: packages/maskbook/src/web3/hooks/useContract.ts [hunk 18]
```diff
@@ -33,6 +34,7 @@ function createContract<T extends Contract>(from: string, address: string, ABI:
                                 from,
                                 to: contract.options.address,
                                 data: cached.encodeABI(),
+                                gasPrice: currentGasPriceSettings.value,
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/hooks/useContract.ts:37]: The initial amount of `currentGasPriceSettings` is zero. You should make sure that you have already set it properly. Maybe a run time reader function is better. Then you can handle complex logic in that.

File: packages/maskbook/src/web3/UI/EthereumGasButton.tsx [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+    const { gasPrices = [], onClose, onSubmit, usd } = props
+    const [gasPrice, setGasPrice] = useState<GasPrice | null>(gasPrices && gasPrices.length > 0 ? gasPrices[0] : null)
+    const [selected, setSeleted] = useState(0)
+
+    const onChange = (index: number, gasPrice: GasPrice) => {
+        setGasPrice(gasPrice)
+        setSeleted(index)
+    }
+    const handleClick = useCallback(() => {
+        onSubmit?.(gasPrice!)
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/UI/EthereumGasButton.tsx:268]: It looks like the `non-null assertion operator` is not necessary.

File: packages/maskbook/src/web3/UI/EthereumGasButton.tsx [hunk 20]
```diff
@@ -0,0 +1,371 @@
...
+        root: {
+            width: '100%',
+            justifyContent: 'left',
+        },
+        content: {
+            width: '100%',
+            display: 'flex',
+            flexDirection: 'column',
+            padding: theme.spacing(1),
+            border: '1px solid grey',
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/UI/EthereumGasButton.tsx:39]: Please use `theme.palette.divider` intead.

File: packages/maskbook/src/web3/UI/EthereumGasButton.tsx (hunk 19 above)
Comment (lelenei) [reply packages/maskbook/src/web3/UI/EthereumGasButton.tsx:268]: if gas price is custom , it  necessary

Comment (guanbinrui) [reply packages/maskbook/src/web3/UI/EthereumGasButton.tsx:268]: But the `onSubmit` can accept a nullish value according to its signature.

File: packages/maskbook/src/web3/constants/index.ts (hunk 13 above)
Comment (lelenei) [reply packages/maskbook/src/web3/constants/index.ts:3]: This is the wait time of custom gas price

File: packages/maskbook/src/web3/apis/gasnow.ts (hunk 15 above)
Comment (lelenei) [reply packages/maskbook/src/web3/apis/gasnow.ts:5]: Interface data is updated every 8s

File: packages/maskbook/src/web3/constants/index.ts (hunk 13 above)
Comment (guanbinrui) [reply packages/maskbook/src/web3/constants/index.ts:3]: 5 minutes or 5 seconds?

**Comment by guanbinrui** [comment]
Hi @lelenei, the commit history looks wired to me. Why you add a775164ee7aa6c14775d7a5d96dc47cef39c7ad1 into it? It's from another PR not related to this one.

File: packages/maskbook/src/web3/constants/index.ts (hunk 13 above)
Comment (lelenei) [reply packages/maskbook/src/web3/constants/index.ts:3]: 5 minutes

**Comment by lelenei** [comment]
> Hi @lelenei, the commit history looks wired to me. Why you add [a775164](https://github.com/DimensionDev/Maskbook/commit/a775164ee7aa6c14775d7a5d96dc47cef39c7ad1) into it? It's from another PR not related to this one.

rebase master

File: packages/maskbook/src/web3/constants/index.ts (hunk 13 above)
Comment (guanbinrui) [reply packages/maskbook/src/web3/constants/index.ts:3]: Please leave a comment.

**Pull request closed without merging**